scripts/lib/setup_venv.py

```diff
@@ -34,6 +35,133 @@
+
+    return index_filename
+
+def get_venv_packages(venv_path):
+    # type: (str) -> Set[str]
+    """
+    Returns the packages installed in the virtual environment using the
+    package index file.
+    """
+    with open(get_index_filename(venv_path)) as reader:
+        return set(p.strip() for p in reader.read().split('\n') if p.strip())
+
+def try_to_copy_venv(venv_path, new_packages):
+    # type: (str, Set[str]) -> bool
+    """
+    Tries to copy packages from an old virtual environment in the cache
+    to the new virtual environment. The algorithm works as follows:
+        1. Find a virtual environment, v, from the cache that has the
+        highest overlap with the new requirements such that:
+            a. The new requirements only add to the packages of v.
+            b. The new requirements only upgrade packages of v.
+        2. Copy the contents of v to the new virtual environment using
+        virtualenv-clone.
+        3. Delete all .pyc files in the new virtual environment.
+    """
+    venv_name = os.path.basename(venv_path)
+
+    overlaps = []  # type: List[Tuple[int, str, Set[str]]]
+    old_packages = set()  # type: Set[str]
+    for sha1sum in os.listdir(VENV_CACHE_PATH):
+        curr_venv_path = os.path.join(VENV_CACHE_PATH, sha1sum, venv_name)
+        if (curr_venv_path == venv_path or
+                not os.path.exists(get_index_filename(curr_venv_path))):
+            continue
+
+        old_packages = get_venv_packages(curr_venv_path)
+        if not (old_packages - new_packages):
+            overlap = new_packages & old_packages
+            overlaps.append((len(overlap), curr_venv_path, overlap))
+
+    target_log = get_logfile_name(venv_path)
+    source_venv_path = None
+    if overlaps:
+        overlaps = sorted(overlaps)
+        _, source_venv_path, copied_packages = overlaps[-1]
+        print('Copying packages from {}'.format(source_venv_path))
+        clone_ve = "{}/bin/virtualenv-clone".format(source_venv_path)
+        cmd = "sudo {exe} {source} {target}".format(exe=clone_ve,
+                                                    source=source_venv_path,
+                                                    target=venv_path).split()
+        try:
+            run(cmd)
+        except Exception:
+            # Virtualenv-clone is not installed. Install it and try running
+            # the command again.
+            run("{}/bin/pip install --no-deps virtualenv-clone".format(
+                source_venv_path).split())
+            run(cmd)
+
+        run(["sudo", "chown", "-R",
+             "{}:{}".format(os.getuid(), os.getgid()), venv_path])
+        run('find {} -name "*.pyc" -delete'.format(venv_path).split())
+        source_log = get_logfile_name(source_venv_path)
+        copy_parent_log(source_log, target_log)
+        create_log_entry(target_log, source_venv_path, copied_packages,
+                         new_packages - old_packages)
+        return True
```

Comment:
Actually, no, the current code will only copy packages from parent venv when all of its packages are present in the child venv. This line, https://github.com/zulip/zulip/pull/1598/files#diff-6c2a6ccaab46a73c6440a20ac3a416f3R106, makes sure of it. So we only add or upgrade packages in child venv.